File: native-android/app/src/test/java/com/upirakshak/engine/RulesEngineTest.kt
package com.upirakshak.engine

import org.junit.Assert.assertEquals
import org.junit.Test

/**
 * Unit tests for RulesEngine
 * Tests all 8 critical scam detection scenarios
 */
class RulesEngineTest {

    @Test
    fun testElectricityScamIsHighRisk() {
        val result = RulesEngine.analyze(
            "WhatsApp",
            "URGENT: Electricity disconnected tonight, scan QR to pay bsescare@icici"
        )
        assertEquals(ThreatLevel.HIGH, result.level)
    }

    @Test
    fun testBankKycScamIsHighRisk() {
        val result = RulesEngine.analyze(
            "SMS",
            "Your SBI account blocked, enter UPI PIN immediately"
        )
        assertEquals(ThreatLevel.HIGH, result.level)
    }

    @Test
    fun testPrizeScamIsHighRisk() {
        val result = RulesEngine.analyze(
            "SMS",
            "Congratulations! Pay ₹500 to receive ₹10,000"
        )
        assertEquals(ThreatLevel.HIGH, result.level)
    }

    @Test
    fun testOtpMessageIsSafe() {
        val result = RulesEngine.analyze(
            "HDFC Bank",
            "Your OTP is 123456. Do not share."
        )
        assertEquals(ThreatLevel.SAFE, result.level)
    }

    @Test
    fun testFriendChatIsSafe() {
        val result = RulesEngine.analyze(
            "WhatsApp",
            "Hi, this is your friend, sending money"
        )
        assertEquals(ThreatLevel.SAFE, result.level)
    }

    @Test
    fun testPaytmCareLookalikeIsMediumRisk() {
        val result = RulesEngine.analyze(
            "SMS",
            "Paytm care: verify your KYC now"
        )
        assertEquals(ThreatLevel.MEDIUM, result.level)
    }

    @Test
    fun testJobOfferScamIsHighRisk() {
        val result = RulesEngine.analyze(
            "WhatsApp",
            "Job offer! Registration fee ₹500 required"
        )
        assertEquals(ThreatLevel.HIGH, result.level)
    }

    @Test
    fun testGovernmentSchemeWithSuspiciousUrlIsHighRisk() {
        val result = RulesEngine.analyze(
            "SMS",
            "Government scheme approved. Click verify-kyc.online"
        )
        assertEquals(ThreatLevel.HIGH, result.level)
    }

    @Test
    fun testRunTestsReturnsAllPassing() {
        val results = RulesEngine.runTests()
        assertEquals(8, results.size)
        results.forEach { (description, passed) ->
            assertEquals("Test failed: $description", true, passed)
        }
    }
}

File: src/lib/rulesEngine.test.ts
import { describe, expect, it } from 'vitest';
import { analyzeMessage } from './rulesEngine';

describe('analyzeMessage', () => {
  it('flags an electricity disconnection scam as HIGH', () => {
    const result = analyzeMessage(
      'WhatsApp',
      'URGENT: Your electricity will be disconnected tonight! Pay now via QR to avoid ₹5000 fine. UPI: bsescare@icici',
    );

    expect(result.level).toBe('HIGH');
    expect(result.reasons.length).toBeGreaterThan(0);
  });

  it('flags a bank-block PIN trap as HIGH', () => {
    const result = analyzeMessage(
      'SMS',
      'Your SBI account has been blocked. Enter UPI PIN to verify and unblock immediately.',
    );

    expect(result.level).toBe('HIGH');
  });

  it('flags a prize-processing-fee scam as HIGH', () => {
    const result = analyzeMessage(
      'SMS',
      'Congratulations! Pay ₹500 processing fee to receive ₹10,000 lottery prize. UPI: refundcare@paytm',
    );

    expect(result.level).toBe('HIGH');
  });

  it('leaves a friendly message alone', () => {
    const result = analyzeMessage('WhatsApp', 'Hi, this is your friend, sending money');

    expect(result.level).toBe('SAFE');
  });

  it('recognises Hinglish urgency keywords', () => {
    const result = analyzeMessage('SMS', 'Aapka account bandh ho jayega, turant verify karein');

    expect(result.matchedPatterns.length).toBeGreaterThan(0);
    expect(result.level).not.toBe('SAFE');
  });

  it('returns a suggested action for every verdict', () => {
    const result = analyzeMessage('SMS', 'random message');

    expect(result.suggestedAction).toBeTruthy();
    expect(result.reasons.length).toBeGreaterThan(0);
  });
});
